pkg/storaged/nfs-mounts.py

```diff
@@ -0,0 +1,344 @@
+        pos = 0
+        while pos < len(buf):
+            (wd, mask, cookie, name_len) = struct.unpack('iIII', buf[pos:pos+16])
+            pos += 16
+            (name,) = struct.unpack('%ds' % name_len, buf[pos:pos + name_len])
+            pos += name_len
+            callback(wd, mask, name.decode().rstrip('\0'))
+
+class Watcher:
+    def __init__(self, path):
+        self.inotify = Inotify()
+        self.path = path
+        self.setup()
+
+    def setup(self):
+        self.wd = self.inotify.add_watch(self.path, IN_CLOSE_WRITE)
+
+    def process(self, callback = None):
+        def event(wd, mask, name):
+            if callback:
+                callback()
+            if mask & IN_IGNORED:
+                self.setup()
+        self.inotify.process(event)
+
+def field_escape(str):
+    return str.replace("\\", "\\134").replace(" ", "\\040").replace("\t", "\\011")
+
+def field_unescape(str):
+    return re.sub("\\\\([0-7]{1,3})", lambda m: chr(int(m.group(1), 8)), str)
+
+def parse_tab(name):
+    entries = [ ]
+    for line in open(name, "r").read().split("\n"):
+        sline = line.strip()
+        if sline == "" or sline[0] == "#":
+            continue
+        fields = list(map(field_unescape, re.split("[ \t]+", sline)))
+        if len(fields) > 0 and ":" in fields[0]:
+            entries.append(fields)
+    return entries
+
+def index_tab(tab):
+    by_remote = { }
+    for t in tab:
+        if not t[0] in by_remote:
+            by_remote[t[0]] = [ ]
+        by_remote[t[0]].append(t)
+    return by_remote
+
+def modify_tab(name, modify):
+    lines = open(name, "r").read().split("\n")
+    if len(lines) > 0 and lines[len(lines)-1] == "":
+        del lines[len(lines)-1]
+
+    new_lines = [ ]
+    for line in lines:
+        sline = line.strip()
+        if sline == "" or sline[0] == "#":
+            new_lines.append(line)
+        else:
+            fields = list(map(field_unescape, re.split("[ \t]+", sline)))
+            if len(fields) > 0 and ":" in fields[0]:
+                new_fields = modify(fields)
+                if new_fields:
+                    if new_fields == fields:
+                        new_lines.append(line)
+                    else:
+                        new_lines.append(" ".join(map(field_escape, new_fields)))
+            else:
+                new_lines.append(line)
+    new_fields = modify(None)
+    if new_fields:
+        new_lines.append(" ".join(map(field_escape, new_fields)))
+
+    f = open(name + ".tmp", "w")
+    f.write("\n".join(new_lines) + "\n")
+    f.flush()
+    os.fsync(f.fileno())
+    f.close()
+    os.rename(name + ".tmp", name)
+
+fstab = [ ]
+fstab_by_remote = { }
+
+mtab = [ ]
+mtab_by_remote = { }
+
+def process_fstab():
+    global fstab, fstab_by_remote
+    fstab = parse_tab("/etc/fstab")
+    fstab_by_remote = index_tab(fstab)
+
+def process_mtab():
+    global mtab, mtab_by_remote
+    mtab = parse_tab("/etc/mtab")
+    mtab_by_remote = index_tab(mtab)
```

Comment:
`/etc/mtab` is really a broken-by-design concept and was abandoned years ago. Let's please not introduce new code for it. `/proc/self/mounts` should do fine.

Reply:
Okay, let's use `/proc/self/mounts`.  The change notification depends on the file being a special file in `/proc`, so the old-school regular `/etc/mtab` file wouldn't work anyway.

(Out of interest, what is broken-by-design re /etc/mtab? Mount namespaces are kinda recent, no?)

Reply:
What's broken is that `/etc/mtab` contains information about all mounted file systems, but resides *on a mounted file system*. I. e. it had become stale in all sorts of situations when mounting stuff while the root fs was still read-only, doesn't work with permanent read-only root scenarios, and is just a layering violation by design.

Reply:
Make sense, thanks!